unblob/extractors/unromfs.py

```diff
@@ -0,0 +1,398 @@
+        target = inode.content.decode("utf-8").lstrip("/")
+
+        if target.startswith(".."):
+            target_path = extract_root.joinpath(output_path.parent, target).resolve()
+        else:
+            target_path = extract_root.joinpath(target).resolve()
+
+        if not is_safe_path(extract_root, target_path):
+            logger.warn(
+                "Path traversal attempt through symlink.", target_path=target_path
+            )
+            return
+        # we create relative paths to make the output directory portable
+        output_path.symlink_to(os.path.relpath(target_path, start=output_path.parent))
+
+    def create_hardlink(self, extract_root: Path, output_path: Path, inode: FileHeader):
+
+        if inode.spec_info in self.inodes:
+            target = str(self.inodes[inode.spec_info].path).lstrip("/")
+            target_path = extract_root.joinpath(target).resolve()
+            # we don't need to check for potential traversal given that, if the inode
+            # is in self.inodes, it already got verified in create_inode.
+            try:
+                os.link(target_path, output_path)
+            except FileNotFoundError:
+                logger.warn(
+                    "Hard link target does not exist, discarding.",
+                    target_path=target_path,
+                )
+        else:
+            logger.warn("Invalid hard link target")
+
+    def create_inode(self, extract_root: Path, inode: FileHeader):
+
+        output_path = extract_root.joinpath(inode.path).resolve()
+        if not is_safe_path(extract_root, inode.path):
+            logger.warn("Path traversal attempt, discarding.", output_path=output_path)
+            return
+        logger.info("dumping inode", inode=inode, output_path=str(output_path))
+
+        if inode.type == FS_TYPE.HARD_LINK:
+            self.create_hardlink(extract_root, output_path, inode)
+        elif inode.type == FS_TYPE.SYMLINK:
+            self.create_symlink(extract_root, output_path, inode)
+        elif inode.type == FS_TYPE.DIRECTORY:
+            output_path.mkdir(mode=inode.mode, exist_ok=True)
+        elif inode.type == FS_TYPE.FILE:
+            with output_path.open("wb") as f:
+                f.write(inode.content)
+        elif inode.type in [FS_TYPE.BLOCK_DEV, FS_TYPE.CHAR_DEV]:
+            os.mknod(inode.path, inode.mode, inode.dev)
+        elif inode.type == FS_TYPE.FIFO:
+            os.mkfifo(output_path, inode.mode)
+
+    def dump_fs(self):
+        # first we create files and directories
+        fd_inodes = {
+            k: v
+            for k, v in self.inodes.items()
+            if v.type in [FS_TYPE.FILE, FS_TYPE.DIRECTORY, FS_TYPE.FIFO, FS_TYPE.SOCKET]
+        }
+        for inode in sorted(fd_inodes.values(), key=lambda inode: inode.path):
+            self.create_inode(self.extract_root, inode)
+
+        # then we create links
+        links_inodes = {
```

Comment:
Cannot a harlink target be a character device? If so, what happens if we are running as a restricted user?

Reply:
Good point. A hard link can point to a character or block device, but the process needs root privileges to create the link.

We should go through char/block devices creation before the links creation. Then, in `create_hardlink`, we should check whether or not the target requires privileges to be linked to. A warning should be emitted if current privileges are not sufficient instead of raising an exception during `os.link()`

Reply:
Switched the order between link and device creation. I moved to imply capturing the PermissionError when unromfs attempts to create a hardlink pointing to a block/char device instead of adding pointless `if` conditions before.